File: solutions/0110. Balanced Binary Tree/0110.cs
// Time complexity: O(n)
// Space complexity: O(h)
public class Solution {
    public bool IsBalanced(TreeNode root) {
        return GetHeight(root) != -1;
    }

    private int GetHeight(TreeNode node) {
        if (node == null) {
            return 0;
        }  
        int leftH = GetHeight(node.left);
        if (leftH == -1) {
            return -1;
        }
        int rightH = GetHeight(node.right);
        if (rightH == -1) {
            return -1;
        }
        if (Math.Abs(leftH - rightH) > 1) {
            return -1;
        } else {
            return Math.Max(leftH, rightH) + 1;
        } 
    }
}

File: solutions/0110. Balanced Binary Tree/0110.js
// Time complexity: O(n)
// Space complexity: O(h)

var isBalanced = function(root) {
    return getHeight(root) !== -1;
}

function getHeight(node) {
    if (node === null) {
        return 0;
    }
    const leftH = getHeight(node.left);
    if (leftH === -1) {
        return -1;
    }
    const rightH = getHeight(node.right);
    if (rightH === -1) {
        return -1;
    }
    if (Math.abs(leftH - rightH) > 1) {
        return -1;
    } else {
        return Math.max(leftH, rightH) + 1;
    }
}